firedrake/preconditioners/hiptmair.py

```diff
@@ -0,0 +1,185 @@
+        pcmg.setOperators(A=A, P=P)
+
+        coarse_solver = pcmg.getMGCoarseSolve()
+        coarse_solver.setOperators(A=coarse_opmat, P=coarse_opmat)
+        # coarse space dm
+        coarse_space = coarse_operator.arguments()[-1].function_space()
+        coarse_dm = coarse_space.dm
+        coarse_solver.setDM(coarse_dm)
+        coarse_solver.setDMActive(False)
+        pcmg.setDM(pc.getDM())
+        pcmg.setFromOptions()
+        self.pc = pcmg
+        self._dm = coarse_dm
+
+        prefix = coarse_solver.getOptionsPrefix()
+        # Create new appctx
+        self._ctx_ref = self.new_snes_ctx(pc,
+                                          coarse_operator,
+                                          coarse_space_bcs,
+                                          coarse_mat_type,
+                                          fcp,
+                                          options_prefix=prefix)
+
+        with dmhooks.add_hooks(coarse_dm, self,
+                               appctx=self._ctx_ref,
+                               save=False):
+            coarse_solver.setFromOptions()
+
+    def update(self, pc):
+        self._assemble_coarse_op()
+        self.pc.setUp()
+
+    def apply(self, pc, X, Y):
+        dm = self._dm
+        with dmhooks.add_hooks(dm, self, appctx=self._ctx_ref):
+            self.pc.apply(X, Y)
+
+    def applyTranspose(self, pc, X, Y):
+        dm = self._dm
+        with dmhooks.add_hooks(dm, self, appctx=self._ctx_ref):
+            self.pc.applyTranspose(X, Y)
+
+    def view(self, pc, viewer=None):
+        super(TwoLevelPC, self).view(pc, viewer)
+        if hasattr(self, "pc"):
+            viewer.printfASCII("Two level PC\n")
+            self.pc.view(viewer)
+
+
+class HiptmairPC(TwoLevelPC):
+
+    _prefix = "hiptmair_"
+
+    def coarsen(self, pc):
+        from firedrake_citations import Citations
+        from firedrake import FunctionSpace, TestFunction, TrialFunction
+        from firedrake.interpolation import Interpolator
+        from ufl.algorithms.ad import expand_derivatives
+        from ufl import (FiniteElement, TensorElement, FacetElement,
+                         replace, zero, grad, curl, as_vector)
+
+        Citations().register("Hiptmair1998")
+        appctx = self.get_appctx(pc)
+        V = dmhooks.get_function_space(pc.getDM())
+        ctx = dmhooks.get_appctx(pc.getDM())
+        problem = ctx._problem
+        a = problem.Jp or problem.J
+        bcs = problem.bcs
+
+        mesh = V.mesh()
+        element = V.ufl_element()
+        formdegree = V.finat_element.formdegree
+        if formdegree == 1:
+            dminus = grad
+            cfamily = "Lagrange"
+            G_callback = appctx.get("get_gradient", None)
+        elif formdegree == 2:
+            dminus = curl
+            if V.shape:
+                dminus = lambda u: as_vector([curl(u[k, ...]) for k in range(u.ufl_shape[0])])
+            cfamily = "N1curl" if mesh.ufl_cell().is_simplex() else "NCE"
```

Comment:
What if you're using elements of the second kind for the H(div) space? Then I expect you should use elements of the second kind for the H(curl) space?